src/lib/zork/zork.ts

```diff
@@ -0,0 +1,378 @@
+    socksSession.onDataForSocksClient((bytes) => {
+      // When too much is buffered, the channel closes/fails.
+      // TODO: backpressure!
+      const BUFFTHRESHOLD = 16000000;  // 16 megabytes
+      if (channel.bufferedAmount < BUFFTHRESHOLD) {
+        channel.send(bytes);
+      } else {
+        console.warn('channel congested, dropping bytes')
+      }
+    });
+    socksSession.onDisconnect(() => {
+      console.info(`[get] ${sessionId}: socks session disconnected`);
+    });
+    channel.onclose = () => {
+      console.info(`[get] ${sessionId}: channel closed (giver side)`);
+    };
+  };
+};
+
+const cmdHandlerByCmdId: {[cmdId: string]: (ctx: Context, cmd?: ParsedCmd) => void} = {
+  'ping': handleCmdPing,
+  'xyzzy': handleCmdXyzzy,
+  'version': handleCmdVersion,
+  'quit': handleCmdQuit,
+  'getters': handleCmdGetters,
+  'transform': handleCmdTransform,
+  'give': handleCmdGive,
+  'get': handleCmdGet
+};
+
+const parseCmd = (cmdline: string) : ParsedCmd => {
+  const tokens = cmdline.split(/\W+/);
+  const cmdId = tokens[0].toLowerCase();
+  const parsed = {id: cmdId, tokens: tokens, source: cmdline};
+  return parsed;
+};
+
+
+// Message handlers:
+
+const handleMsgForModeGive = (ctx: Context, msg: string) => {
+  const parsed = JSON.parse(msg);
+  if (parsed.sdp) {
+    console.info(`[give] got sdp`);//: ${msg}`);
+    ctx.peerConn.setRemoteDescription(parsed);
+  } else {
+    console.info(`[give] parsed.sdp missing, ignoring msg`);//: ${msg}`);
+  }
+};
+
+const handleMsgForModeGet = (ctx: Context, msg: string) => {
+  const parsed = JSON.parse(msg);
+  if (parsed.type === 'icecandidate') {
+    console.info(`[get] icecandidate from client ${ctx.clientId}`);//: ${msg}`);
+    ctx.peerConn.addIceCandidate(parsed.candidate);
+  } else if (parsed.type === 'offer') {
+    console.info(`[get] got offer`);//: ${msg}`);
+    ctx.peerConn.setRemoteDescription(parsed);
+    ctx.peerConn.createAnswer((answer: any) => {
+      const answerJson = JSON.stringify(answer);
+      console.info(`[get] created answer`);//: ${answerJson}`);
+      ctx.peerConn.setLocalDescription(answer);
+      ctx.socket.write(`${JSON.stringify(answer)}\n`);
+    }, console.error);
+  } else {
+    console.error(`[get] unexpected msg: ${msg}`);
+  }
+};
+
+const msgHandlerByMode: {[mode: string]: (ctx: Context, msg: string) => void} = {
+  'give': handleMsgForModeGive,
+  'get': handleMsgForModeGet
+};
+
+const handleMsg = (ctx: Context, msg: string) => {
+  if (ctx.mode) {
+    // Already in 'give' or 'get' mode. Dispatch to corresponding message handler.
+    const msgHandler = msgHandlerByMode[ctx.mode];
+    if (msgHandler) {
+      msgHandler(ctx, msg);
+    } else {
+      console.error(`no message handler for mode: ${ctx.mode}`);
+    }
+  } else {
+    // Not yet in 'give' or 'get' mode. Treat msg as command.
+    const cmd = parseCmd(msg);
+    const cmdHandler = cmdHandlerByCmdId[cmd.id] || handleCmdInvalid;
+    cmdHandler(ctx, cmd);
+  }
+};
```

Comment:
In the earlier version of this which we went over last week, I had an extra bit of state `shouldKeepHandlingCommands` (akin to the `keepParsing` flag in freedom-based zork), and the command handlers would return true or false to indicate whether to continue parsing input as commands.

But that state is immediately derivable based on whether the `mode` flag has been set yet (e.g. to "give" or "get"). So I took out that extra flag to simplify and reduce noise in the code, in exchange for making the logic a little less explicit: rather than all command handlers returning true or false to indicate the new state, now the two that would flip the flag to false instead just set the `mode` flag, which they'd have to do anyway.

It seems simple enough that, along with the accompanying explanatory comments, I think this is preferable, but happy to make any suggested changes.